openssl/src/crypto/signature.rs

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2016, The rust-openssl developers.
+
+//! Interface for computing/verifying message signatures
+//!
+//! The `Signer` allows for the computation of crytographic signatures
+//! of data given a private key.  The corresponding public key can then
+//! be used to verify the integrity and authenticity of signed messages
+//! using the `Verifier`.
+//!
+//! # Example
+//!
+//! ```rust
+//! use std::io::prelude::*;
+//! use openssl::crypto::signature::{Signer, Verifier};
+//! use openssl::crypto::pkey::PKey;
+//! use openssl::crypto::hash;
+//!
+//! // Generate public/private keypairs
+//! let mut keypair = PKey::new();
+//! let mut privkey = PKey::new();
+//! let mut pubkey = PKey::new();
+//! keypair.gen(1024);
+//! privkey.load_priv(&keypair.save_priv()[..]);
+//! pubkey.load_pub(&keypair.save_pub()[..]);
+//!
+//! // generate some data
+//! let data1: Vec<u8> = (0..25).cycle().take(1024).collect();
+//! let data2: Vec<u8> = (100..150).cycle().take(1024 * 5).collect();
+//!
+//! // sign some content with private key
+//! let mut signer = Signer::new(hash::Type::SHA512, &privkey).unwrap();
+//! signer.write_all(&data1[..]).unwrap();
+//! signer.write_all(&data2[..]).unwrap();
+//! let signature = signer.finish().unwrap();
+//!
+//! // verify content using signature and public key
+//! let mut verifier = Verifier::new(hash::Type::SHA512, &pubkey).unwrap();
+//! verifier.write_all(&data1[..]).unwrap();
+//! verifier.write_all(&data2[..]).unwrap();
+//! assert!(verifier.finish(&signature[..]).unwrap());
+//! ```
+
+use std::ptr;
+use std::io::{self, Write};
+use std::marker::PhantomData;
+use libc;
+
+use ffi;
+use ssl::error::SslError;
+use crypto::pkey::PKey;
+use crypto::hash;
+
+/// Signature Generation Context
+pub struct Signer<'a> {
+    ctx: *mut ffi::EVP_MD_CTX,
+    pkey: PhantomData<&'a PKey>,
+}
+
+/// Signature Verification Context
+pub struct Verifier<'a> {
+    ctx: *mut ffi::EVP_MD_CTX,
+    pkey: PhantomData<&'a PKey>,
+}
+
+impl<'a> Signer<'a> {
+    /// Create and initialize a new Signer
+    ///
+    /// The digest type specified will determine the final form  of the
+    /// signature returned by `finalize`.
+    pub fn new(digest_type: hash::Type, pkey: &'a PKey) -> Result<Signer, SslError> {
+        ffi::init();
+
+        let ctx = unsafe {
+            let r = ffi::EVP_MD_CTX_create();
+            assert!(!r.is_null(), "EVP_MD_CTX_create failed");
+            r
+        };
+
+        let md = digest_type.evp_md();
+
+        unsafe {
+            try_ssl_if!(1 !=
```

Comment:
Ah right. It might make sense to update `try_ssl` to `$e != 1` instead of `$e == 0`.